client.go

```diff
@@ -1329,31 +1357,40 @@ func (c *HostClient) Do(req *Request, resp *Response) error {
 		if hasBodyStream {
 			break
 		}
+		// Path prioritization based on ease of computation
+		attempts++
 
-		switch {
-		case c.RetryIf != nil:
-			retry = c.RetryIf(req)
-		case c.RetryIfErr != nil:
-			retry = c.RetryIfErr(req, err)
-		default:
-			retry = isIdempotent(req)
+		if attempts >= maxAttempts {
+			break
+		}
+		if c.RetryIfErr != nil {
+			sleepDuration, resetReqTimeout, retry = c.RetryIfErr(req, attempts, err)
+		} else {
+			retry = retryFunc(req)
 		}
-
 		if !retry {
+			break
 			// Retry non-idempotent requests if the server closes
 			// the connection before sending the response.
 			//
 			// This case is possible if the server closes the idle
 			// keep-alive connection on timeout.
 			//
 			// Apache and nginx usually do this.
-			if err != io.EOF {
-				break
-			}
+			//
+			// This assertion cannot ensure that the server has not read our request.
+			// Even if part of the request has been read, retrying for a non-idempotent
+			// request is still unsafe. It is better to leave this decision to the user.
+			//
+			// if err != io.EOF {
+			//	break
+			// }
```

Comment:
I agree, lets just remove the comment to make the code cleaner.

Reply:
ok